plan/planbuilder.go

```diff
@@ -951,154 +935,251 @@ func (b *planBuilder) buildInsert(insert *ast.InsertStmt) Plan {
+	exprCols := make([]*expression.Column, 0, len(insert.Setlist))
 	for _, assign := range insert.Setlist {
-		col, err := insertPlan.tableSchema.FindColumn(assign.Column)
+		exprCol, err := insertPlan.tableSchema.FindColumn(assign.Column)
 		if err != nil {
 			b.err = errors.Trace(err)
-			return nil
+			return
 		}
-		if col == nil {
+		if exprCol == nil {
 			b.err = errors.Errorf("Can't find column %s", assign.Column)
-			return nil
+			return
 		}
+		colNames = append(colNames, assign.Column.Name.L)
+		exprCols = append(exprCols, exprCol)
+	}
 
-		// Check whether the column to be updated is the generated column.
-		if columnByName[assign.Column.Name.L].IsGenerated() {
-			b.err = ErrBadGeneratedColumn.GenByArgs(assign.Column.Name.O, tableInfo.Name.O)
-			return nil
+	// Check whether the column to be updated is the generated column.
+	tCols, err := table.FindCols(insertPlan.Table.Cols(), colNames, tableInfo.PKIsHandle)
+	if err != nil {
+		b.err = errors.Trace(err)
+		return
+	}
+	for _, tCol := range tCols {
+		if tCol.IsGenerated() {
+			b.err = ErrBadGeneratedColumn.GenByArgs(tCol.Name.O, tableInfo.Name.O)
+			return
 		}
+	}
 
+	for i, assign := range insert.Setlist {
 		expr, _, err := b.rewriteWithPreprocess(assign.Expr, mockTablePlan, nil, true, checkRefColumn)
 		if err != nil {
 			b.err = errors.Trace(err)
-			return nil
 		}
 		insertPlan.SetList = append(insertPlan.SetList, &expression.Assignment{
-			Col:  col,
+			Col:  exprCols[i],
 			Expr: expr,
 		})
 	}
-
 	insertPlan.Schema4OnDuplicate = insertPlan.tableSchema
-	if insert.Select != nil {
-		selectPlan := b.build(insert.Select)
-		if b.err != nil {
-			return nil
-		}
+}
+
+func (b *planBuilder) buildValuesListOfInsert(insert *ast.InsertStmt, insertPlan *Insert, mockTablePlan *LogicalTableDual,
+	checkRefColumn func(n ast.Node) ast.Node) {
+	affectedValuesCols, err := insertPlan.getAffectCols(insert)
+	if err != nil {
+		b.err = errors.Trace(err)
+		return
+	}
 
-		numInsertCols := mathutil.Min(selectPlan.Schema().Len(), len(tableInfo.Columns))
-		// If the column to be inserted in the insert table is a generated
-		// column, raises a "ErrBadGeneratedColumn" error here.
-		for _, col := range tableInfo.Columns[:numInsertCols] {
+	if len(insert.Columns) > 0 || len(insert.Lists[0]) > 0 {
+		// If value_list is not empty or the col_list is not empty, length of value_list should be the same with col_list's.
+		if len(insert.Lists[0]) != len(affectedValuesCols) {
+			b.err = ErrWrongValueCountOnRow.GenByArgs(1)
+			return
+		}
+		// No generated column is allowed.
+		for _, col := range affectedValuesCols {
 			if col.IsGenerated() {
-				b.err = ErrBadGeneratedColumn.GenByArgs(col.Name.O, tableInfo.Name.O)
-				return nil
+				b.err = ErrBadGeneratedColumn.GenByArgs(col.Name.O, insertPlan.Table.Meta().Name.O)
+				return
 			}
 		}
+		// If the value_list and col_list is empty and we have generated column, we can still write to this table.
```

Comment:
This comment should reside to `if len(insert.Columns) == 0 && len(insert.Lists[0]) == 0` branch? Maybe we can remove it?

Reply:
Yes, it's for the `else` branch, explaining that else branch doesn't need to do anything. 
I think it's useful. But maybe we can move it other place.

Reply:
I think we can put this comment to the `if len(insert.Columns) > 0 || len(insert.Lists[0]) > 0 {` to explain why we don't check whether there is generated column in the table when the `value_list` and `col_list` is empty.